internal/resources/kubernetessecret/data_source_secret.go

```diff
@@ -17,60 +18,74 @@ import (
 	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/kubernetessecret/spec"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/kubernetessecret/status"
 )
 
+type dataFromServer struct {
+	UID                     string
+	secretExportRespNil     bool
+	secretExportErr         error
+	meta                    *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta
+	atomicSpec              *secretclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceSecretSpec
+	clusterScopeStatus      *secretclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceSecretStatus
+	clusterGroupScopeStatus *secretclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceSecretStatus
+}
+
 func DataSourceSecret() *schema.Resource {
 	return &schema.Resource{
 		ReadContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-			return dataSourceSecretRead(context.WithValue(ctx, contextMethodKey{}, DataSourceRead), d, m)
+			return dataSourceSecretRead(helper.GetContextWithCaller(ctx, helper.DataRead), d, m)
 		},
 		Schema: getDataSourceSchema(),
 	}
 }
 
-func dataSourceSecretRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+func dataSourceSecretRead(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
 	config := m.(authctx.TanzuContext)
 
-	var (
-		diags            diag.Diagnostics
-		secretResp       *secretclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretResponse
-		secretexportResp *secretexportclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretExportResponse
-		err              error
-	)
+	secretName, ok := d.Get(NameKey).(string)
+	if !ok {
+		return diag.Errorf("Unable to read secret name")
+	}
+
+	secretNamespaceName, ok := d.Get(NamespaceNameKey).(string)
+	if !ok {
+		return diag.Errorf("Unable to read secret namespace name")
+	}
 
-	secretName, _ := d.Get(NameKey).(string)
+	scopedFullnameData := scope.ConstructScope(d, secretName, secretNamespaceName)
 
-	secretResp, err = config.TMCConnection.SecretResourceService.SecretResourceServiceGet(constructFullname(d))
-	if err != nil || secretResp == nil {
-		if clienterrors.IsNotFoundError(err) {
-			if ctx.Value(contextMethodKey{}) == DataSourceRead {
-				return diag.FromErr(errors.Wrapf(err, "Tanzu Mission Control cluster secret entry not found, name : %s", d.Get(NameKey)))
-			}
+	if scopedFullnameData == nil {
+		return diag.Errorf("Unable to get Tanzu Mission Control secret entry; Scope full name is empty")
+	}
 
+	secretDataFromServer, err := retrieveSecretDataFromServer(config, scopedFullnameData, d)
+	if err != nil {
+		if clienterrors.IsNotFoundError(err) && !helper.IsDataRead(ctx) {
 			_ = schema.RemoveFromState(d, m)
-
-			return diags
+			return
 		}
 
-		return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret entry, name : %s", secretName))
```

Comment:
Good to have the detailed error message

Reply:
I've added detailed error msg to `retrieveSecretDataFromServer` method since now k8s secret have 2 scopes, PTAL